nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -304,4 +329,174 @@ WaypointFollower::goalResponseCallback(
+  for (auto waypoint : waypoint_follower_goal_.poses) {
+    RCLCPP_DEBUG(
+      this->get_logger(),
+      "\t(%lf, %lf)", waypoint.pose.position.x, waypoint.pose.position.y);
+  }
+
+  auto goal_options = ActionClientGPS::SendGoalOptions();
+  goal_options.result_callback = std::bind(
+    &WaypointFollower::GPSResultCallback, this,
+    std::placeholders::_1);
+  goal_options.goal_response_callback = std::bind(
+    &WaypointFollower::GPSGoalResponseCallback, this,
+    std::placeholders::_1);
+
+  auto future_goal_handle = waypoint_follower_action_client_->async_send_goal(
+    waypoint_follower_goal_, goal_options);
+  if (rclcpp::spin_until_future_complete(
+      waypoint_follower_client_node_,
+      future_goal_handle) != rclcpp::FutureReturnCode::SUCCESS)
+  {
+    RCLCPP_ERROR(get_logger(), "Send goal call failed");
+    return;
+  }
+  // Get the goal handle and save so that we can check
+  // on completion in the timer callback
+  waypoint_follower_goal_handle_ = future_goal_handle.get();
+  if (!waypoint_follower_goal_handle_) {
+    RCLCPP_ERROR(get_logger(), "Goal was rejected by server");
+    return;
+  }
+}
+
+std::vector<geometry_msgs::msg::PoseStamped>
+WaypointFollower::convertGPSWaypointstoPosesinMap(
+  const
+  std::vector<sensor_msgs::msg::NavSatFix> & gps_waypoints,
+  const rclcpp_lifecycle::LifecycleNode::SharedPtr & parent_node,
+  const rclcpp::Client<robot_localization::srv::FromLL>::SharedPtr & fromll_client)
+{
+  RCLCPP_INFO(parent_node->get_logger(), "Converting GPS waypoints to Map Frame..");
+  if (!parent_node) {
+    throw std::runtime_error{
+            "Failed to lock node while trying to convert GPS waypoints to Map frame"};
+  }
+  std::vector<geometry_msgs::msg::PoseStamped> poses_in_map_frame_vector;
+  int index_of_gps_waypoints = 0;
+  for (auto && curr_gps_waypoint : gps_waypoints) {
+    auto fromLLRequest = std::make_shared<robot_localization::srv::FromLL::Request>();
+    fromLLRequest->ll_point.latitude = curr_gps_waypoint.latitude;
+    fromLLRequest->ll_point.longitude = curr_gps_waypoint.longitude;
+    fromLLRequest->ll_point.altitude = curr_gps_waypoint.altitude;
+    if (!fromll_client->wait_for_service((std::chrono::seconds(10)))) {
+      RCLCPP_ERROR(
+        parent_node->get_logger(),
+        "fromLL service from robot_localization is not available"
+        "cannot convert GPS waypoints to Map frame poses,"
+        "Make sure you have run navsat_transform_node of robot_localization");
+      return std::vector<geometry_msgs::msg::PoseStamped>();
+    }
+    RCLCPP_INFO(
+      parent_node->get_logger(), "Processing for %i'th Waypoint..",
+      index_of_gps_waypoints);
+    auto inner_client_callback =
+      [&, parent_node](rclcpp::Client<robot_localization::srv::FromLL>::SharedFuture inner_future)
+      {
+        auto result = inner_future.get();
+      };
+    auto inner_future_result = fromll_client->async_send_request(
```

Comment:
https://github.com/ros-planning/navigation2/blob/main/nav2_util/include/nav2_util/service_client.hpp using this service client will make this code more compact and straight forward